Autoloader: Support unoptimized PSR-4

ObliviousHarmony opened the pull request:
#### Changes proposed in this Pull Request:

One of the things that the autoloader does when generating the manifest is convert all PSR-4 classes into a classmap. While this is great for performance it can be somewhat error-prone for those who are used to the standard operation of Composer's autoloader. As I've started adding more new classes I've found myself constantly forgetting to dump the autoloader!

This PR proposes utilizing the `-o` option that Composer provides in order to only do this conversion in production builds. We do this by generating a manifest with the namespace paths and versions and then loading via PSR-4 when either a classmap option is unavailable or is of an older version. Once I began working on this change I started to see places that I could refactor to increase test granularity as well as clean up the code a bit. I'll highlight some of the new classes and their purpose as well as the changes I've made to existing classes.

- **Notable Changes**
  - `AutoloadGenerator`
    - Now generates a `jetpack_autoload_psr4.php` manifest containing all of the namespace paths and their corresponding versions when ran using `composer dump-autoload` or `composer install` without the `-o` option.
    - I moved the copied autoloader class files from `vendor/` to `vendor/jetpack-autoloader` so that they are more easily identified. This also would make the rename in #16755 unnecessary! I examined how this would play with existing autoloaders and it would do the same as the referenced PR.
- **New Classes**
  - `Manifest_Handler`
    - After adding a `Psr4_Handler` and copying the `Classes_Handler` I recognized that all of these classes are virtually identical. I then opted to refactor them into a single class and write tests that should cover all three cases.
  - `Version_Loader`
    - With the addition of PSR-4 autoloading and the need for version reconciliation there I felt it would make sense if the autoloading itself took place in its own class.

One scenario I wanted to make sure was adequately handled was if one plugin was using classmap loading while another was using PSR-4 loading. This is a situation that is very present in WooCommerce development and it was important that it be addressed. This is handled in the `Version_Loader` during which it will check both sources and load the class from the newest source.

**_One potentially significant difference highlighted in testing this with WooCommerce is the lack of support for invalid class namespaces when autoloading via PSR-4._** Since no `$autoloadType` is passed to `ClassMapGenerator::createMap()` we've avoiding even [seeing this issue](https://github.com/composer/composer/issues/8403), and if you run an optimized dump, Composer will complain even though our `AutoloadGenerator` will still spit out those files anyway. I'm not completely sure yet how I feel about this one, as again, these classes would not have loaded if Composer was used instead. 

_I think_ users can just move/rename files and they'll work correctly in a backwards-compatible way. My leaning here is that perhaps we should be passing the `$autoloadType` as-expected and be consistent with Composer here. Developers can then note the error when they update the autoloader and fix the classes that are incorrect.

#### Jetpack product discussion
N/A

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:

* Set up a WooCommerce development environment cloned to `master`
* Replace the `vendor/automattic/jetpack-autoloader/src` folder with the contents from this PR
* Generate the autoloader using `composer dump-autoloader` and make sure `phpunit tests/php/src` passes
* Generate the autoloader using `composer dump-autoloader -o` and make those test still pass

**(Note that the unoptimized autoloader will give an error but pass the tests, this is because of an invalid class namespace)**

#### Proposed changelog entry for your changes:
* Autoloader: Added support for non-optimized PSR-4 namespace loading.

ObliviousHarmony committed:
Fixed incorrect autoloader generated filenames

ObliviousHarmony committed:
Refactored the `Classes_Handler` and `Files_Handler` into a generic class

As I started adding a `Psr4_Handler` I realized all of these classes are essentially identical and opted to refactor them.

ObliviousHarmony committed:
Added non-optimized mode for using PSR-4 autoloading during development

ObliviousHarmony committed:
Corrected the mistaken directory separator at the end of PSR-4 paths

ObliviousHarmony committed:
Fixed version reconciliation for `AutoloadGenerator` in non-optimized builds

ObliviousHarmony committed:
Moved generated autoloader files into dedicated folder

As this refactor continues to add new classes it becomes apparent these might benefit from having a new home.

ObliviousHarmony committed:
Moved generated autoloader files into dedicated folder

As this refactor continues to add new classes it becomes apparent these might benefit from having a new home.

ObliviousHarmony committed:
Refactored the autoloader into a dedicated class

Since PSR-4 and classmap files need to have their version compared as well, I took the time to refactor that into a new class.

ObliviousHarmony committed:
Added support for multiple directories per PSR-4 namespace

ObliviousHarmony committed:
Optimized the PSR-4 autoloading

kbrown9 reviewed (changes requested):
I reviewed the code changes and left a few comments in-line. I still need to do manual testing. Thank you for the refactors, especially for cleaning up the file generation code and combining the Classes_Handler and Files_Handler classes!

ObliviousHarmony commented on packages/autoloader/src/AutoloadGenerator.php:
```diff
@@ -193,17 +198,65 @@ protected function parseAutoloadsType( array $packageMap, $type, PackageInterfac
+	 */
+	private function getPsr4Map( array $autoloads, $scanPsrPackages, Filesystem $filesystem, $vendorPath, $basePath ) {
+		if ( $scanPsrPackages ) {
+			return 'array();' . PHP_EOL;
```
That's a good point and would also solve your other comment about not checking the PSR-4 file for the autoloader first since it would't pass the `file_exists` check.

kbrown9 replied:
Is there a reason to check the PSR-4 file for the autoloader first? Since the autoloader generator path should be in the classmap file in production, it seems like we should check the classmap first even if we don't generate the PSR-4 file.

ObliviousHarmony replied:
One of the things I was thinking about when writing this is that the `jetpack_autoload_psr4.php` file will be much smaller than the `jetpack_autoload_classmap.php` file. If we're saying the PSR-4 file is only generated when used then we can be confident that if it exists at all then the autoloader will be there and can avoid loading the classmap file unnecessarily.

Another thing I was thinking about when working on this is that we could probably do a bit of optimization around the manifest loading and the autoloader version reconciliation. We're currently loading the same manifest twice and I wonder if perhaps we should be storing the autoloader version using some different mechanism given the potentially large filesizes of manifest files for plugins with many dependencies.

kbrown9 replied:
> We're currently loading the same manifest twice and I wonder if perhaps we should be storing the autoloader version using some different mechanism given the potentially large filesizes of manifest files for plugins with many dependencies.

One option would be to create a small file that returns an array containing just the autoloader version the the value of `$scanPsrPackages`. During the autoloader version reconciliation, we could determine the latest installed autoloader version and whether any plugin on the site is using a PSR-4 classmap. Were you thinking of something like that, or were you thinking about a different approach?

ObliviousHarmony replied:
That's exactly what I was thinking 😄 There do seem to be some other benefits too in that we can write information about the state of the autoloader at built-time and avoid some `file_exists` checks too.

ObliviousHarmony replied:
I've been thinking a bit about this and I think plugin load order might make it problematic. If a plugin with the `v2.0.2` autoloader initializes first then it won't know to check this file in the first place and will expect it in the classmap. I'm going to inject `Automattic\Jetpack\Autoloader\AutoloadGenerator` into the classmap to deal with this case and given that I think we can just rely on the lack of PSR-4 file presence as an indicator that the autoload is optimized.

ObliviousHarmony commented on packages/autoloader/src/functions.php:
```diff
@@ -42,16 +51,43 @@ function autoloader( $class_name ) {
+	$manifest_handler->register_plugin_manifests(
+		'vendor/composer/jetpack_autoload_psr4.php',
+		$jetpack_packages_psr4
+	);
```
If we don't generate the PSR-4 file in optimized builds we could use the autoloader version comparison to also detect if an autoloader is in an unoptimized state to skip this manifest check, what do you think?

kbrown9 replied:
Yeah, I think that makes sense.

ObliviousHarmony replied:
Since the `Manifest_Handler` does an `is_readable()` check before actually loading the manifest we probably can't do much more than we already are. One option might be to add a constant that can be used for optimization to entirely disable PSR-4 autoloading?

I've also adjusted the hander to only look up the active plugin paths once.

kbrown9 commented:
> One potentially significant difference highlighted in testing this with WooCommerce is the lack of support for invalid class namespaces when autoloading via PSR-4. 

I don't have a really strong opinion about this. I lean towards leaving it the way it is, because I don't know if any of the plugins that use the autoloader will need to fix incorrect classes before updating the autoloader. Also, I don't want to force plugins to make changes that could cause other problems, for example, during an update. Do you know if the Woo plugins have any package classes that don't comply with the PSR-4 standard?

ObliviousHarmony committed:
Refactored the `AutoloadGenerator` into several classes to add unit tests

ObliviousHarmony committed:
Updated the autoloader in Jetpack to correct composer plugin namespace

ObliviousHarmony committed:
Ensured the `AutoloadGenerator` is always written to the classmap

I've also moved all of the Composer classes back to the root namespace to make it easier to deal with backwards compatibility. This change will make it easier to compare versions between 2.0 autoloaders without getting strange.

ObliviousHarmony committed:
Corrected the autoloader plugin

ObliviousHarmony committed:
Merged the lookups for active plugins paths to increase performance

ObliviousHarmony commented:
> Do you know if the Woo plugins have any package classes that don't comply with the PSR-4 standard?

When I run `composer dump-autoloader -o` against WooCommerce I receive the following deprecation notices:
```
Deprecation Notice: Class WP_CLI located in ./vendor/wp-cli/wp-cli/php/class-wp-cli.php does not comply with psr-0 autoloading standard. It will not autoload anymore in Composer v2.0.
Deprecation Notice: Class WP_CLI_Command located in ./vendor/wp-cli/wp-cli/php/class-wp-cli-command.php does not comply with psr-0 autoloading standard. It will not autoload anymore in Composer v2.0.
Deprecation Notice: Class Automattic\WooCommerce\Blocks\StoreApi\Utilities\ProductQueryFilters located in ./packages/woocommerce-blocks/src/StoreApi/Utilities/ProductFiltering.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0.
Deprecation Notice: Class Automattic\WooCommerce\Admin\Composer\Package located in ./packages/woocommerce-admin/src/Package.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0.
```

In the Blocks case the file is named incorrectly and it doesn't look like renaming it will cause any update problems either (cc @nerrad)
In the Admin case this looks to have been done to solve a bug caused by the plugin not previously requiring `autoload_packages.php` and instead using `autoload.php`. It should be fine to fix this too.

> I don't want to force plugins to make changes that could cause other problems, for example, during an update.

That's a fair point as there may be other plugins depending on the incorrectly namespaced classes. I'm not entirely sure there are any cases like that in the wild, as the above classes are internal to their respective plugins and wouldn't have any issues being used 

--

My feeling is that this is something we should fix now before it becomes more of an issue down the line. Not passing an `$autoloadType` to `ClassMapGenerator::createMap()` accidentally causes classes to be loaded would otherwise not have been. If someone was to switch from the Jetpack autoloader to Composer directly for instance they might run into problems that aren't clear when classes don't load that used to.

I do think it's probably important that we be as consistent with the behavior of Composer as possible, especially given that it will spit out deprecation notices (or errors depending on version) when they run a `composer dump-autoload -o` build even with our current setup.

As far as developer solutions go, technically they can also add the file to a `classmap` declaration in their `composer.json` file. This would cause it to always be included in the classmap despite what warnings and errors may indicate. The preference of course would be to fix the class file but if that might cause a problem on update in their specific case they can avoid it.

That said however, one thing that I'm wary of is introducing something that, if accidentally released to production, will cause a plugin to not find classes it could before. Another option would be for us to _always_ generate the classmap for namespaces in unoptimized builds. Personally I'd like to avoid this as it feels more like a hack for something we should fix before it becomes a more noticeable problem. I wonder if perhaps this might be something we should define with a constant for developers to decide, what do you think?

ObliviousHarmony committed:
Fixed erroneous slashes after using `trailingslashit`

ObliviousHarmony committed:
Updated Jetpack lock file with new autoload directive for the autoloader package

ObliviousHarmony committed:
Simplified the manifest generator a little bit

ObliviousHarmony committed:
Made some of the autoloader unit tests less granular to reduce fragility

ObliviousHarmony committed:
Refactored the `Autoloader_Handler` to minimize the number of plugin lookups

ObliviousHarmony committed:
Integrated CodeClimate feedback

ObliviousHarmony committed:
Fixed a bug with filemap loading

ObliviousHarmony committed:
Integrated CodeClimate feedback

ObliviousHarmony committed:
Fixed the name of the filemap manifest

ObliviousHarmony committed:
Restructured the autoloader test suite and included all of the missing dummy manifest data

ObliviousHarmony committed:
Added test for filemap loading

ObliviousHarmony committed:
Added a cleanup to remove all of the legacy autoloader files

ObliviousHarmony committed:
[not verified] Adjusted autoloader gitignore to include all manifests

Note: Pushed without verification because the plugin paths aren't liked by phpcs-changed!

ObliviousHarmony committed:
Added integration tests verifying that parsed manifests can be autoloaded

kbrown9 commented on packages/autoloader/src/AutoloadProcessor.php:
```diff
@@ -0,0 +1,162 @@
+			return null;
+		}
+
+		$classBlacklist = null;
```
What do you think of naming this something like `$excludeList`? We've been working on removing terms like 'blacklist', as described in [this issue](https://github.com/Automattic/jetpack/issues/16099).

ObliviousHarmony replied:
I've gone with `$excludedClasses` to keep what its excluding clear, thanks!

kbrown9 commented on packages/autoloader/src/class-version-loader.php:
```diff
@@ -0,0 +1,156 @@
+	public function find_class_file( $class_name ) {
+		$data = $this->find_newest_file(
+			isset( $this->classmap[ $class_name ] ) ? $this->classmap[ $class_name ] : null,
+			$this->find_psr4_file( $class_name )
```
What do you think of checking whether `$this->psr4_map` is set here in `find_class_file`? If it's not set, we can return the file path from the classmap (if it exists), and we wouldn't need to call `find_newest_file()` or `find_psr4_file()`. Since the PSR-4 map won't be used in production, the package class file paths will be obtained from the classmap most of the time.

kbrown9 replied:
We discussed this and decided not to make any changes because they would reduce code readability.

kbrown9 reviewed (commented):
I've looked through the code, and I just have a few comments! 

I still have to do some manual testing with different plugin versions and multiple plugins.

kbrown9 commented:
> My feeling is that this is something we should fix now before it becomes more of an issue down the line. Not passing an `$autoloadType` to `ClassMapGenerator::createMap()` accidentally causes classes to be loaded would otherwise not have been. If someone was to switch from the Jetpack autoloader to Composer directly for instance they might run into problems that aren't clear when classes don't load that used to.

I guess I would prefer that we accidentally load the classes over having a fatal occur :smile: Jetpack uses classmap autoloading, so I don't have a strong opinion either way. We could open a PR, get the consumer plugins to test it out, and see if it causes any problems.
 
> That said however, one thing that I'm wary of is introducing something that, if accidentally released to production, will cause a plugin to not find classes it could before. Another option would be for us to _always_ generate the classmap for namespaces in unoptimized builds. Personally I'd like to avoid this as it feels more like a hack for something we should fix before it becomes a more noticeable problem. I wonder if perhaps this might be something we should define with a constant for developers to decide, what do you think?

I'm not sure why a developer would want to always generate a classmap for unoptimized builds or have a constant that would allow them to do that. They should probably fix the class file, or as you suggested, add the file to a classmap declaration in their composer.json file if the class file can't be fixed.

zinigor commented:
So I have tried to catch up with y'all here, and I think I have succeeded. Overall I like where this is going, but I agree with Kim that we should make sure consumers sanity check their setups before introducing safeguards that would be a bit over the top. Not reporting the issues is a valid concern, though, maybe we can think of some kind of linter based on Composer's libraries?
Also, it looks like this PR still conflicts with something and some tests are failing, we need to look into that.

kbrown9 commented on packages/autoloader/src/AutoloadGenerator.php:
```diff
@@ -73,56 +80,26 @@ public function dump(
+		// Write all of the files now that we're done.
+		$this->writeAutoloaderFiles( $vendorPath . '/jetpack-autoloader/', $suffix );
+		$this->writeManifests( $vendorPath . '/' . $targetDir, $processedAutoloads );
 	}
```
Do you think there would be any value in displaying a warning when an unoptimized autoloader is generated? Something like this:

```
if ( ! $scanPsrPackages ) {
    $this->io->writeError( "<warning>Generating an unoptimized autoloader. If this is a production build, consider using the -o flag to generate an optimized autoloader.</warning>" );
}
```

ObliviousHarmony replied:
I was actually planning to add this today as I'd like to avoid introducing silent problems for those who update to this next version of the autoloader. I know for instance that in our build scripts we use `composer install` and so we're going to need to address that.

kbrown9 commented:
@ObliviousHarmony - I left a few comments requesting minor changes in the code. If you'd like, I can push commits for those changes.

kbrown9 commented on packages/autoloader/src/functions.php:
```diff
@@ -103,8 +78,8 @@ function set_up_autoloader() {
  */
-function reset_maps_after_update( $response, $hook_extra, $result ) {
-	global $jetpack_packages_classmap;
+function reset_maps_after_update( $response, $hook_extra, $result ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
```
We can remove this.

ObliviousHarmony commented:
I've rebased and force-pushed to get all of the other adjustments that have been made to the autoloader since this was started!

ObliviousHarmony committed:
Applied changes according to feedback

* Added a warning when generating unoptimized builds
* Renamed the `$classBlacklist`

ObliviousHarmony commented:
I'll acquiesce that changing the existing behavior is outside the scope of this change and we can revisit that in another PR with proper testing to avoid holding this up. Thanks for all of the feedback!

kbrown9 reviewed (approved):
I tested this alongside other versions of the autoloader (v1 and v2), and the package classes and files were loaded as expected. I also did a quick test on Windows, and everything was fine.

@zinigor - I think this is ready to merge. There are a few items which need separate PRs:
- Handling invalid class namespaces when autoloading via PSR-4.
- Detecting situations where both v1 and v2 autoloaders files are present (we started that in PR #16962).

zinigor reviewed (approved):
Thanks for working on this @ObliviousHarmony ! Let's merge!
